### PR TITLE
Cache and gen

### PR DESCRIPTION
**Pull request by anshradh**
This has working implementations of sampling (with various modifications to logits, i.e. temperature, top-k, top-p, etc.) and greedy decoding, both with and without caching past keys and values (cache offers 3-4x speedup). These are also all batched. Will put up a PR for beam search next

### CODE REVIEW

**Comment by neelnanda-io** [comment]
Thanks and sorry for taking so long on this! Did a bunch of refactoring and integration into my new changes, and just merged in :)

**Comment by anshradh** [comment]
No worries and sorry for not keeping it more up-to-date!
